test/provider/ChannelHandler.unittest.ts

```diff
@@ -0,0 +1,260 @@
+import 'reflect-metadata';
+
+import {ChannelHandler} from '../../src/provider/controller/ChannelHandler';
+import {Model} from '../../src/provider/model/Model';
+import {Signal1} from '../../src/provider/common/Signal';
+import {AppWindow} from '../../src/provider/model/AppWindow';
+import {DesktopContextChannel, ContextChannel} from '../../src/provider/model/ContextChannel';
+import {ChannelError, Context, FDC3ChannelEventType} from '../../src/client/main';
+import {createMockChannel, createMockAppWindow} from '../mocks';
+
+jest.mock('../../src/provider/model/Model');
+
+const mockOnChannelChanged = jest.fn<void, [AppWindow, ContextChannel | null, ContextChannel | null]>();
+
+let mockModel: Model;
+
+let channelHandler: ChannelHandler;
+
+beforeEach(() => {
+    jest.resetAllMocks();
+
+    mockModel = new Model(null!, null!, null!);
+    (mockModel as any)['onWindowAdded'] = new Signal1<AppWindow>();
+    (mockModel as any)['onWindowRemoved'] = new Signal1<AppWindow>();
+
+    channelHandler = new ChannelHandler(mockModel);
+    channelHandler.onChannelChanged.add((appWindow: AppWindow, channel: ContextChannel | null, previousChannel: ContextChannel | null) => {
+        mockOnChannelChanged(appWindow, channel, previousChannel);
+    });
+});
+
+it('When getting desktop channels, ChannelHandler only returns desktop channels', () => {
+    const testChannels = [
+        {...createMockChannel(), id: 'test-1'},
+        {...createMockChannel(), id: 'test-2'},
+        {...createMockChannel(), id: 'test-3'}
+    ];
+
+    (testChannels[0] as any)['type'] = 'desktop';
+    (testChannels[2] as any)['type'] = 'desktop';
+
+    (mockModel as any)['channels'] = testChannels;
+
+    expect(channelHandler.getDesktopChannels()).toEqual([testChannels[0], testChannels[2]]);
+});
+
+describe('When geting channel by ID', () => {
+    it('If Model returns a channel, ChannelHandler returns the channel', () => {
+        const testChannel = new DesktopContextChannel('test', 'test', 0);
+
+        (mockModel.getChannel as jest.Mock<ContextChannel | null, [string]>).mockReturnValue(testChannel);
+
+        expect(channelHandler.getChannelById('test')).toEqual(testChannel);
+    });
+
+    it('If Model returns null, ChannelHandler throws an exception', () => {
+        (mockModel.getChannel as jest.Mock<ContextChannel | null, [string]>).mockReturnValue(null);
```

Comment:
A lot of this casting seems unnecessary, there's Jest types that I think can clean this up (will need a cast when you create the mock, but then not when using it).

Can we discuss when you have the time?

Reply:
Fixed. Notes:
- jest.spyOn doesn't seem to work as expected (throws an exception), maybe because the object I'm applying it to is already a mock, so using Object.definePropery
- Is there a way to make `Writable` (defined int `types.ts`) apply only to a single property?

Reply:
This should do it:
`export type PartiallyWritable<T, P extends keyof T> = Writable<Pick<T, P>> & Omit<T, P>;`

Use like:
`let mockModel: jest.Mocked<PartiallyWritable<Model, 'onWindowAdded' | 'onWindowRemoved'>>;`

It'll take a certain subset of properties of `Writable<T>`, and combine that with the opposite subset of properties of `T`.

Reply:
👍